### ISSUE TITLE
Ненадёжно работает сохранение треков localStorage

### ISSUE THREAD

**Issue by wladich** [wladich/nakarte#228]
Симптомы:
1.  Редкие жалобы отдельных пользователей на пропадание треков при перезагрузке браузера
2. В логах пишутся события открытия и закрытия сайта. Закрытие совмещено с сохранением треков в localStorage.  Для примерно одной трети событий открытия сессии нет события закрытия. Для Айфонов -- 100%. Возможные причины: a) не срабатывает beforeunload b) не шлётся запрос из обработчика события beforeunload. Чтобы понять, надо попробовать сохранять  события закрытия в localStorage и при очередном открытии сайта их отправлять.

Другая проблемa: сейчас сохранение сделано в виде стека --  трек из последней закрытой вкладки будет открыт в первой открытой. Это работает только для некоторых сценариев -- перезагрузка одной вкладки,  закрытие одной-двух вкладок и восстановление по Ctrl-Shift-T.  Плохо работает при перезагрузке браузера, если открыто несколько вкладок  с nakarte, треки загружаются не в те вкладки, в которых были до перезагрузки. Кроме того, браузер при старте не загружает сразу все вкладки, а только одну видимую, остальные загружает когда пользователь на них переключается. Возможен такой сценарий:
Открыта вкладка с треками. Перезагружаем браузер (вкладка с треками не активна). Открываем ещё одну вкладку с nakarte, закрываем (в стек запишется состояние без треков). Переключаемся на старую вкладку. Только теперь сайт загружается и из стека достаётся пустое состояние -- пользователь видит, что трек пропал. На самом деле трек в стеке ещё есть и если открыть ещё одну вкладку, он загрузится, но это явно неправильное поведение.

**Comment by wladich** [comment]
Идеи по новой систем сохранения:
1) Сохранять регулярно, например каждые 10 секунд, плюс в beforeunload.
2) Сохранять по уникальному ID вкладки -- но как его получить? Он должен переживать перезагрузку браузера.
3) Проверить, выживает ли sessionStorage перезагрузку, восстановление вкладки и  перезагрузку браузера.
4) Посмотреть, выживает ли содержимое textarea в форме.

**Comment by myadzel** [comment]
К сожалению, по опыту работы с beforeunload, он не дает гарантии на какие-то действия, особенно при работе с файловой системой. 

Идентификатор генерировать и записывать в sessionStorage, но получение записанного значения после восстановления работает вроде только в Chrome (как и восстановление значения в textarea).

**Comment by slazav** [comment]
Мне всегда кажется, что сохранение в localstorage - это непредсказуемо и ненадежно. Я в одной вкладке нарисовал один трек, в другой - другой, потом что-то закрыл, где-то открыл nakarte по ссылке. И я не могу вспомнить, что у меня будет на экране при следующей загрузке. Довольно неуютно открывать сайт, не зная, что на нем будет нарисовано (и в любом случае, обычно все это приходится очищать)...

Я бы подумал, как из сохранения в localstorage сделать удобный, но не обязательный инструмент. Кнопки "сохранить/восстановить данные локально", меню "история" или что-то подобное.

**Comment by wladich** [comment]
Так таск именно про это - сохранение работает не так, как планировалось, из-за архитектурного про...счета. Я сам от этого особо не страдал, т.к. много открытых вкладок не держу, а пользователи сильно не жаловались. 

Для определённости: главное (и пока единственное)  назначение данной функции: во вкладках должны сохраняться треки:
 - при перезагрузке вкладки
 - при закрытии вкладки и восстановлении по Ctrl-Z/Ctrl-Shift-T
 - при закрытии (в т.ч. аварийном) и октрытии браузера с несколькими вкладками

Возможность восстановления треков из истории -- это задача для светлого будущего.

**Comment by wladich** [comment]
Ещё одна потенциальная проблема с существующей версии: при старте делается очистка старых записей, но нет никакой синхронизации между потоками, т.е. несколько одновременно стартовавших вкладок могут начать чистить хранилище вместе.

**Comment by wladich** [comment]
Всё должно починиться в #1258. Выложил на https://stage.nakarte.me, хорошо бы потестировать по мере возможности.

**Comment by wladich** [comment]
Однострочник для дампа и восстановления содержимого localStorage - полезно для тестирования импорта треков из старого хранилища

https://gist.github.com/wladich/3e08887ad2bbbf36f7333426781eefb7

**Comment by slazav** [comment]
А можешь написать, изменилось ли старая логика, или должно починиться там, где не работало?

**Comment by wladich** [comment]
Да, логика поменялась и поведение должно стать более предсказуемым.

Изменение поведения:
- При открытии новой страницы или ссылки без треков старые треки неожиданно не загружаются, всегда пустой список.
- При восстановлении закрытой вкладки всё надёжно загружается, причём нет старого ограничения в 10 вкладок, можно до 100.
- сохранение делается при изменении треков, а не при закрытии вкладки, т.е. треки сохраняются даже при аварийном завершении браузера
- также нормально сохраняются треки, когда браузер выгружает приложение в неактивной вкладке, а потом загружает заново
- при закрытии+открытии браузера с несколькими вкладками треки загружаются в правильные табы, а не куда придётся.


Изменения в логике:
- странное решение со стеком сессий убрал
- Теперь при старте генерируется рандомный ID сессии и хранится в `window.history`. Если при загрузке в `window.history` уже есть sessionId, то используется он (это для случаев перезагрузки страницы, восстановления закрытой вкладки, засыпания вкладки)
- треки сохраняются по ключу sessionId
- треки пишутся: при старте, при изменении, при завершении (последнее может быть ненадёжно)

Ещё добавился контрол со списком активных и неактивных сессий, из него можно переключать вкладки и открывать старые сессии

**Comment by wladich** [comment]
Пока нашёл одну неоднозначную ситуацию:
- Открываем две вкладки: N1 и N2 в N2 рисуем трек.
- Закрываем N2
- В N1 отркываем список сессий, видим вверху списка неактивную сессию из N2. Кликаем по ней, появляется окно с треками из N2
- Нажимаем Ctrl-Shift-T/Ctrl-Z - восстанавливается оригинальная вклада N2, позиция карты и слои на ней восстановились, но треков нет - их "забрала" вкладка открытая через список сессий

**Comment by slazav** [comment]
Ага, заметил, что слева появилась новая кнопка!

Но, может, кнопку надо назвать не "Menu", а "Sessions"? Или туда что-то еще хочется спрятать?

И, может быть, у каждой сессии сделать кнопки "открыть" и "удалить"? 

В целом я такое изменение одобряю, буду пользоваться.

**Comment by wladich** [comment]
Да, я думал туда ещё чего-нибудь добавить, начал с копии функкции "скопировать ссылку на треки".

Про удалить - тоже думал, посмотрим, насколько оно надо. Там же начнутся случайные удаления, просьбы сделать корзину или отмену...
